### ISSUE TITLE
Push GUI should honor git's push.default setting (especially 'current')

### ISSUE THREAD

**Issue by johndog** [gitextensions/gitextensions#7118]
## Current behaviour

Git Ex will always default to showing the upstream branch as the push target, regardless of the value of the push.default setting.

## Expected behaviour

Git Extensions GUI should honor push.default, and present a default that matches.

## Steps to reproduce

- Initially configure git so that "git config --get push.default" returns "current"
- Run "git checkout origin/master your-topic-branch", so that origin/master is an upstream.
- Run "git push"... git will push to origin/your-topic-branch, no origin/master.
- In Git Extensions, click the 'Push' button to bring up the GUI
- The UI will default to pushing your-topic-branch to origin/master.

Expected: because of the "current" setting, it should default to pushing to a branch of the same name.

## Screenshots

![image](https://user-images.githubusercontent.com/22213780/64560666-c7134b00-d2fd-11e9-96e2-504a3cbd4b83.png)

## Did this work in previous version of GitExtensions
No

## Environment

- Git Extensions 3.1.1.6049
- Build 2f872105b4eb42421d9fdafd68e90a9e27d20d96
- Git 2.22.0.vfs.1.1.57.gbaf16c8
- Microsoft Windows NT 10.0.18970.0
- .NET Framework 4.8.3921.0
- DPI 96dpi (no scaling)

**Comment by johndog** [comment]
Note:  Honoring 'simple' would be more difficult, as the UI would have to provide an error that the upstream doesn't match.  It might be better to just treat 'simple' as 'current' for the purposes of UI.

**Comment by PhiLhoSoft** [comment]
I totally agree with this issue (I came here for that), and I disagree on one point: it did work in previous versions of GitExtensions, so it is a regression bug.

I have been bitten a couple of times: I am used to work on a feature / fix branch (eg. `branch_name`), and to click the Commit & push button, which pushed to `origin/branch_name`, asking if we want to create it if it doesn't exist yet. Now, it pushes to `origin/master`, which is certainly not the intended goal! (Cannot make a merge request, and so on.)

**Comment by gerhardol** [comment]
The setting in 2.x was also separate, so the behavior is not changed in 3.x from what I see.
If someone has a better suggestion for how the default tracked branch is derived, the code is here:
https://github.com/gitextensions/gitextensions/blob/master/GitUI/CommandsDialogs/FormPush.cs#L729

**Comment by RussKie** [comment]
> it did work in previous versions of GitExtensions, so it is a regression bug.

No, it isn't so.
The behaviour is consistent to what we had as far as v2.5x (and likely before that).

You need to check an association between the local and a remote branches. If you create a branch from origin/master GE will create a link, so when you decide to push it will honour that link.

You can see the relationships either via _Repository > Remote repositories... > Default pull behavior_, or in the status bar of the Commit dialog on the right.

**Comment by PhiLhoSoft** [comment]
OK. When I look at this tab (never used it before!), indeed I see the problematic branch (the one I worked on when I posted above comment) has a dot (empty?) (in place of origin for other branches) for Remote repository, and a Default merge with set to master instead of the name of the (local) branch.
So, for some reason (likely to be my fault, obviously), this branch wasn't configured like the others, and I falsely blamed the software... 😅
Sorry about that. But at least, this bug is here as reference for people with similar problem...

**Comment by pmiossec** [comment]
@philhosoft could you sum this up in the first message (mostly copy/paste this last message) before closing the issue, please?

**Comment by johndog** [comment]
PhiLhoSofts issue was incorrectly associated with  this issue, please don't close it.  The original issue stands.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by pmiossec** [comment]
OK. Sorry.

**Comment by RussKie** [comment]
> * Initially configure git so that "git config --get push.default" returns "current"

How?

I have it unset:
![image](https://user-images.githubusercontent.com/4403806/65598326-821a3480-dfde-11e9-967b-1a080b078f79.png)

**Comment by johndog** [comment]
> > * Initially configure git so that "git config --get push.default" returns "current"
> 
> How?

I'll add this to the repro:

`git config --global push.default current`

--global is technically unnecessary, but I suspect this is rarely useful as a repo specific option.  The only reason I left it vague is because I meant that what "git config --get" returns is what GitEx should use (regardless of how it is configured).

**Comment by johndog** [comment]
Also added a note:

> Note: Aside from conformance, the main reason I value this issue is that a topic branch configured with 'master' as an upstream but which pushes to 'current' is very elegantly maintained with "git pull" and "git pull --rebase" -- but the push experience with that configuration is currently encumbered in the UI.

**Comment by johndog** [comment]
@gerhardol wrote:

> If someone has a better suggestion for how the default tracked branch is derived...

In case it's not yet clear (not trying to be pedantic), the notion of "default tracked branch" shouldn't actually exist anymore.  You can have a "default push branch", or a "tracked branch", but not a "default tracked branch".  The distinction is that "tracking" is always applied in the pull case, but it takes a back seat to "push.default" in the push case.

Looking at the code in FormPush.cs, it took me a while to develop a useful understanding, in part due to naming:  

`                            string defaultRemote = _remotesManager.GetDefaultPushRemote(_selectedRemote,
                                branch.Name);`

In this line, I've deduced that 'defaultRemote' is actually a branch, not a remote -- in a domain with both 'remotes' and 'branches', using 'remote' as a synonym for 'branch' is a very confusing choice.  Moreover, 'git push' has a default remote, and each remote has a default branch, so having the GetDefaultPushRemote() function return a default branch instead of a default remote is also confusing.  Maybe the function has overloaded purpose or something, preventing it from having a more specific name.

In any case, it seems like _remotesManager.GetDefaultPushRemote[Branch]() would be the most useful function in which to implement the logic to choose the default push branch based on push.default.  Then the FormPush.cs code wouldn't change at all, and anyone else using that function would get a more accurate result (unless they were using it to query the tracking branch... in which case the remotesManager should have a new function to query the Tracking branch separately)
